test/unit/core/metafunction/range_iterator_test.cpp

```diff
@@ -0,0 +1,278 @@
+    if constexpr (pos < list1::size() - 1)
+        assert_same_types<list1, list2, pos + 1>();
+}
+
+TEST(range_and_iterator, value_type_)
+{
+    using foreign_iterator = detail::random_access_iterator<std::vector<int>>;
+    // TODO also add something with proxies
+    auto v = ranges::view::ints(1);
+    using type_list = meta::list<value_type_t<std::vector<int>>,                    // short
+                                 typename value_type<std::vector<int>>::type,       // long
+                                 typename std::vector<int>::value_type,             // member type
+                                 value_type_t<std::vector<int> const>,              // const container
+                                 value_type_t<iterator_t<std::vector<int>>>,        // iterator
+                                 value_type_t<foreign_iterator>,                    // iterator2
+                                 value_type_t<decltype(v)>>;                        // range, no member
+    using comp_list = meta::list<int,
+                                 int,
+                                 int,
+                                 int,
+                                 int,
+                                 int,
+                                 int>;
+
+    assert_same_types<type_list, comp_list>();
+}
+
+TEST(range_and_iterator, reference_)
+{
+    using foreign_iterator = detail::random_access_iterator<std::vector<int>>;
+    auto v = ranges::view::ints(1);
+    using type_list = meta::list<reference_t<std::vector<int>>,                    // short
+                                 typename reference<std::vector<int>>::type,       // long
+                                 typename std::vector<int>::reference,             // member type
+                                 reference_t<std::vector<int> const>,              // const container
+                                 reference_t<iterator_t<std::vector<int>>>,        // iterator
+                                 reference_t<foreign_iterator>,                    // iterator2
+                                 reference_t<decltype(v)>>;                        // range, no member
+
+    using comp_list = meta::list<int &,
+                                 int &,
+                                 int &,
+                                 int const &,   // container is const
+                                 int &,
+                                 int &,
+                                 int>;          // view creates values
+
+    assert_same_types<type_list, comp_list>();
+}
+
+TEST(range_and_iterator, rvalue_reference_)
+{
+    using foreign_iterator = detail::random_access_iterator<std::vector<int>>;
+    auto v = ranges::view::ints(1);
+    using type_list = meta::list<rvalue_reference_t<std::vector<int>>,                    // short
+                                 typename rvalue_reference<std::vector<int>>::type,       // long
+// No types have member,yet:
```

Comment:
Maybe add a todo, that this needs to be tested. Or write a dummy class, that exposes this rvalue_reference type.

Reply:
I am not sure anything is ever supposed to define, currently this is generated in ranges:: also all the time.